### PR TITLE
[PH] TPS Scenario Runner

### PR DESCRIPTION
**Pull request by oschwaldp-oci**
### TPS Scenario Runner
Create script and adapt existing scripts to do a binary search of tps range at low test duration with primitive scoring algorithm to find where Transfer Transactions Per Second (TPS) seems to be topping out.  Then do a longer duration run at the discovered TPS.

Performance Test Results Report reports the results of the initial short duration test binary search to zero in on max TPS as well as the follow-on longer running tests to determine if the max TPS was sustainable and zero in on the true max.

### Arguments
**New** Arguments for control of `performance_test`:

- `--max-tps-to-test` - The max target transfers realistic as ceiling of test range, default=50000
- `--test-iteration-duration-sec` - The duration of transfer trx generation for each iteration of the test during the initial search (seconds), default=30
- `--test-iteration-min-step` - The step size determining granularity of tps result during initial search, default=500
- `--final-iterations-duration-sec` - The duration of transfer trx generation for each final longer run iteration of the test during the final search (seconds), default=90
- `--save-json` - Whether to save overarching performance run report, default=False
- `--save-test-json` - Whether to save json reports from each test scenario, default=False

_Existing_ Arguments to configure pass-through to `performance_test_basic` test scenarios:
- `--tps-limit-per-generator` - Maximum amount of transactions per second a single generator can have., default=4000
- `--genesis` - Path to genesis.json, default="tests/performance_tests/genesis.json"
- `--num-blocks-to-prune` - The number of potentially non-empty blocks, in addition to leading and trailing size 0 blocks, to prune from the beginning and end of the range of blocks of interest for evaluation., default=2
- `-p` - producing nodes count
- `-n` - total nodes
- `-d` - delay between nodes startup
- `-s` - topology
- `--nodes-file` - File containing nodes info in JSON format.
- `--dump-error-details` - Upon error print etc/eosio/node_*/config.ini and var/lib/node_*/stderr.log to stdout
- `-v` - verbose logging
- `--leave-running` - Leave cluster running after test finishes
- `--clean-run` - Kill all nodeos and kleos instances
- `--keep-logs` - Don't delete var/lib/node_* folders, or other test specific log directories, upon test completion

### Report: Performance Test

Command used to run test and generate report:
```
./tests/performance_tests/performance_test.py --test-iteration-duration-sec 10 --final-iterations-duration-sec 30 --save-json True
```

**Note**: Truncated some search result scenarios as well as the bulk of the `MaxTpsReport` and `LongRunningMaxTpsReport` as they are single performance test basic run reports as detailed at the end of this PR.

`InitialMaxTpsAchieved` - the max TPS throughput achieved during initial, short duration test scenarios to narrow search window
`LongRunningMaxTpsAchieved` - the max TPS throughput achieved during final, longer duration test scenarios to zero in on sustainable max TPS

```
{
  "InitialMaxTpsAchieved": 16200,
  "LongRunningMaxTpsAchieved": 15400,
  "InitialSearchResults": {
    "0": {
      "success": false,
      "searchTarget": 25000,
      "searchFloor": 0,
      "searchCeiling": 50000,
      "basicTestResult": {
        "targetTPS": 25000,
        "resultAvgTps": 15382.714285714286,
        "expectedTxns": 250000,
        "resultTxns": 250000,
        "tpsExpectMet": false,
        "trxExpectMet": true,
        "basicTestSuccess": true,
        "logsDir": "performance_test/2022-10-12_16-55-27/testRunLogs/performance_test_basic/2022-10-12_16-55-27"
      }
    },
    "1": {
      "success": true,
      "searchTarget": 12500,
      "searchFloor": 0,
      "searchCeiling": 25000,
      "basicTestResult": {
        "targetTPS": 12500,
        "resultAvgTps": 12499.375,
        "expectedTxns": 125000,
        "resultTxns": 125000,
        "tpsExpectMet": true,
        "trxExpectMet": true,
        "basicTestSuccess": true,
        "logsDir": "performance_test/2022-10-12_16-55-27/testRunLogs/performance_test_basic/2022-10-12_16-57-15"
      }
    },
    "2": {
      "success": false,
      "searchTarget": 18800,
      "searchFloor": 12500,
      "searchCeiling": 25000,
      "basicTestResult": {
        "targetTPS": 18800,
        "resultAvgTps": 16209.105263157895,
        "expectedTxns": 188000,
        "resultTxns": 188000,
        "tpsExpectMet": false,
        "trxExpectMet": true,
        "basicTestSuccess": true,
        "logsDir": "performance_test/2022-10-12_16-55-27/testRunLogs/performance_test_basic/2022-10-12_16-58-53"
      }
    },
    "3": {
      "success": true,
      "searchTarget": 15600,
      "searchFloor": 12500,
      "searchCeiling": 18800,
      "basicTestResult": {
        "targetTPS": 15600,
        "resultAvgTps": 15623.1875,
        "expectedTxns": 156000,
        "resultTxns": 156000,
        "tpsExpectMet": true,
        "trxExpectMet": true,
        "basicTestSuccess": true,
        "logsDir": "performance_test/2022-10-12_16-55-27/testRunLogs/performance_test_basic/2022-10-12_17-00-35"
      }
    },
    "4": {
      "success": false,
      "searchTarget": 17200,
      "searchFloor": 15600,
      "searchCeiling": 18800,
      "basicTestResult": {
        "targetTPS": 17200,
        "resultAvgTps": 16264.64705882353,
        "expectedTxns": 172000,
        "resultTxns": 172000,
        "tpsExpectMet": false,
        "trxExpectMet": true,
        "basicTestSuccess": true,
        "logsDir": "performance_test/2022-10-12_16-55-27/testRunLogs/performance_test_basic/2022-10-12_17-02-15"
      }
    },
    "5": {
      "success": false,
      "searchTarget": 16400,
      "searchFloor": 15600,
      "searchCeiling": 17200,
      "basicTestResult": {
        "targetTPS": 16400,
        "resultAvgTps": 16263.235294117647,
        "expectedTxns": 164000,
        "resultTxns": 164000,
        "tpsExpectMet": false,
        "trxExpectMet": true,
        "basicTestSuccess": true,
        "logsDir": "performance_test/2022-10-12_16-55-27/testRunLogs/performance_test_basic/2022-10-12_17-03-55"
      }
    },
    "6": {
      "success": true,
      "searchTarget": 16000,
      "searchFloor": 15600,
      "searchCeiling": 16400,
      "basicTestResult": {
        "targetTPS": 16000,
        "resultAvgTps": 16098.9375,
        "expectedTxns": 160000,
        "resultTxns": 160000,
        "tpsExpectMet": true,
        "trxExpectMet": true,
        "basicTestSuccess": true,
        "logsDir": "performance_test/2022-10-12_16-55-27/testRunLogs/performance_test_basic/2022-10-12_17-05-36"
      }
    },
    "7": {
      "success": true,
      "searchTarget": 16200,
      "searchFloor": 16000,
      "searchCeiling": 16400,
      "basicTestResult": {
        "targetTPS": 16200,
        "resultAvgTps": 16135.5625,
        "expectedTxns": 162000,
        "resultTxns": 162000,
        "tpsExpectMet": true,
        "trxExpectMet": true,
        "basicTestSuccess": true,
        "logsDir": "performance_test/2022-10-12_16-55-27/testRunLogs/performance_test_basic/2022-10-12_17-07-16"
      }
    }
  },
  "InitialMaxTpsReport": {
    "Analysis": {
<truncated>
    },
    "args": {
<truncated>
    },
<truncated>
  },
  "LongRunningSearchResults": {
    "0": {
      "success": false,
      "searchTarget": 16200,
      "searchFloor": 14700,
      "searchCeiling": 17700,
      "basicTestResult": {
        "targetTPS": 16200,
        "resultAvgTps": 15782.413793103447,
        "expectedTxns": 486000,
        "resultTxns": 486000,
        "tpsExpectMet": false,
        "trxExpectMet": true,
        "basicTestSuccess": true,
        "logsDir": "performance_test/2022-10-12_16-55-27/testRunLogs/performance_test_basic/2022-10-12_17-08-56"
      }
    },
    "1": {
      "success": true,
      "searchTarget": 15400,
      "searchFloor": 14700,
      "searchCeiling": 16200,
      "basicTestResult": {
        "targetTPS": 15400,
        "resultAvgTps": 15343.875,
        "expectedTxns": 462000,
        "resultTxns": 462000,
        "tpsExpectMet": true,
        "trxExpectMet": true,
        "basicTestSuccess": true,
        "logsDir": "performance_test/2022-10-12_16-55-27/testRunLogs/performance_test_basic/2022-10-12_17-11-11"
      }
    },
    "2": {
      "success": false,
      "searchTarget": 15800,
      "searchFloor": 15400,
      "searchCeiling": 16200,
      "basicTestResult": {
        "targetTPS": 15800,
        "resultAvgTps": 15523.30357142857,
        "expectedTxns": 474000,
        "resultTxns": 474000,
        "tpsExpectMet": false,
        "trxExpectMet": true,
        "basicTestSuccess": true,
        "logsDir": "performance_test/2022-10-12_16-55-27/testRunLogs/performance_test_basic/2022-10-12_17-13-24"
      }
    },
    "3": {
      "success": false,
      "searchTarget": 15600,
      "searchFloor": 15400,
      "searchCeiling": 15800,
      "basicTestResult": {
        "targetTPS": 15600,
        "resultAvgTps": 15464.589285714286,
        "expectedTxns": 468000,
        "resultTxns": 468000,
        "tpsExpectMet": false,
        "trxExpectMet": true,
        "basicTestSuccess": true,
        "logsDir": "performance_test/2022-10-12_16-55-27/testRunLogs/performance_test_basic/2022-10-12_17-15-38"
      }
    }
  },
  "LongRunningMaxTpsReport": {
    "Analysis": {
<truncated>
    },
    "args": {
<truncated>
    },
<truncated>
  },
  "args": {
    "killAll": false,
    "dontKill": false,
    "keepLogs": false,
    "dumpErrorDetails": false,
    "delay": 1,
    "nodesFile": null,
    "verbose": false,
    "_killEosInstances": true,
    "_killWallet": true,
    "pnodes": 1,
    "totalNodes": 0,
    "topo": "mesh",
    "extraNodeosArgs": " --http-max-response-time-ms 990000 --disable-subjective-api-billing true ",
    "useBiosBootFile": false,
    "genesisPath": "tests/performance_tests/genesis.json",
    "maximumP2pPerHost": 5000,
    "maximumClients": 0,
    "_totalNodes": 2,
    "testDurationSec": 10,
    "finalDurationSec": 30,
    "maxTpsToTest": 50000,
    "testIterationMinStep": 500,
    "tpsLimitPerGenerator": 4000,
    "saveJsonReport": true,
    "saveTestJsonReports": false,
    "numAddlBlocksToPrune": 2,
    "logsDir": "performance_test/2022-10-12_16-55-27"
  },
  "env": {
    "system": "Linux",
    "os": "posix",
    "release": "5.10.102.1-microsoft-standard-WSL2"
  },
  "nodeosVersion": "v3.2.0-dev"
}
```

### Updates to existing in support of new functionality:
Rework `performance_test_basic.py` to work as an import module and as a script.

This was necessary as a first step to being able to use this functionality as an import module for an orchestrator to run multiple instances of `PerformanceBasicTest` with different tps configurations.

Some rework to how args are handled in the report.  When run as an import module, command line args won't be the same or available.  Get args/config info from how the `PerformanceBasicTest` was configured instead.

Created `TestHelperConfig` and `ClusterConfig` classes to help simplify and separate concerns for all the parameters to the test.

### Report: Performance Test Basic
Output report for individual test run now looks like:
```
{
    "Analysis": {
      "BlockSize": {
        "avg": 1507950.0350877193,
        "emptyBlocks": 0,
        "max": 1897400,
        "min": 1184064,
        "numBlocks": 57,
        "sigma": 140462.7045683851
      },
      "BlocksGuide": {
        "configAddlDropCnt": 2,
        "firstBlockNum": 2,
        "lastBlockNum": 259,
        "leadingEmptyBlocksCnt": 1,
        "setupBlocksCnt": 127,
        "tearDownBlocksCnt": 37,
        "testAnalysisBlockCnt": 57,
        "testEndBlockNum": 222,
        "testStartBlockNum": 129,
        "totalBlocks": 258,
        "trailingEmptyBlocksCnt": 32
      },
      "TPS": {
        "avg": 15343.875,
        "configTestDuration": 30,
        "configTps": 15400,
        "emptyBlocks": 0,
        "max": 17218,
        "min": 13555,
        "numBlocks": 57,
        "sigma": 695.6516488285334
      },
      "TrxCPU": {
        "avg": 43.294225108225106,
        "max": 1389.0,
        "min": 24.0,
        "samples": 462000,
        "sigma": 15.451334956307504
      },
      "TrxLatency": {
        "avg": 0.4002558766164821,
        "max": 0.806999921798706,
        "min": 0.10100007057189941,
        "samples": 462000,
        "sigma": 0.15376674034615292
      },
      "TrxNet": {
        "avg": 24.567108225108225,
        "max": 25.0,
        "min": 24.0,
        "samples": 462000,
        "sigma": 0.4954760197252983
      }
    },
    "args": {
      "_killEosInstances": true,
      "_killWallet": true,
      "_totalNodes": 2,
      "delay": 1,
      "dontKill": false,
      "dumpErrorDetails": false,
      "expectedTransactionsSent": 462000,
      "extraNodeosArgs": " --http-max-response-time-ms 990000 --disable-subjective-api-billing true ",
      "genesisPath": "tests/performance_tests/genesis.json",
      "keepLogs": false,
      "killAll": false,
      "maximumClients": 0,
      "maximumP2pPerHost": 5000,
      "nodesFile": null,
      "numAddlBlocksToPrune": 2,
      "pnodes": 1,
      "saveJsonReport": false,
      "targetTps": 15400,
      "testTrxGenDurationSec": 30,
      "topo": "mesh",
      "totalNodes": 0,
      "tpsLimitPerGenerator": 4000,
      "useBiosBootFile": false,
      "verbose": false
    },
    "completedRun": true,
    "env": {
      "os": "posix",
      "release": "5.10.102.1-microsoft-standard-WSL2",
      "system": "Linux"
    },
    "nodeosVersion": "v3.2.0-dev"
  }
```

### CODE REVIEW

File: tests/performance_tests/performance_test_basic.py [hunk 1]
```diff
@@ -15,218 +15,291 @@
...
+        genesisPath: str = "tests/performance_tests/genesis.json"
+        maximumP2pPerHost: int = 5000
+        maximumClients: int = 0
+        loggingDict = { "bios": "off" }
+        _totalNodes: int = 2
+
+        def __post_init__(self):
+            self._totalNodes = max(2, self.pnodes if self.totalNodes < self.pnodes else self.totalNodes)
+
+    def __init__(self, testHelperConfig: TestHelperConfig=TestHelperConfig(), clusterConfig: ClusterConfig=ClusterConfig(), targetTps: int=8000, testTrxGenDurationSec: int=30, tpsLimitPerGenerator: int=4000, numAddlBlocksToPrune: int=2, rootLogDir: str=os.path.splitext(os.path.basename(__file__))[0], saveJsonReport: bool=False):
```
Comment (jgiszczak) [inline tests/performance_tests/performance_test_basic.py:51]: There comes a point when the code is begging for a carriage return...

Comment (oschwaldp-oci) [reply tests/performance_tests/performance_test_basic.py:51]: Never!  There comes a point when the code is begging for a bigger monitor.  Point taken.  I hadn't done any sort of final formatting sweep.  I'll take care of this.  Thanks.

Comment (oschwaldp-oci) [reply tests/performance_tests/performance_test_basic.py]: Addressed: [11179c7](https://github.com/AntelopeIO/leap/pull/302/commits/11179c764269316ce52d8f6d099530d805a0d9d4)

Comment (jgiszczak) [reply tests/performance_tests/performance_test_basic.py]: I bought my first UltraHD monitor in 2014.  :stuck_out_tongue:  I am aware that other people still don't have one, and Github intentionally doesn't insert linewraps the way a sophisticated editor does, so I do appreciate a few carriage returns.

**Review by jgiszczak — APPROVED** [review]

File: tests/performance_tests/performance_test.py [hunk 2]
```diff
@@ -0,0 +1,271 @@
+#!/usr/bin/env python3
+
+from dataclasses import dataclass, asdict, field
+import os
+from platform import release, system
+import sys
+import json
+from datetime import datetime
+import shutil
```
Comment (ClaytonCalabrese) [inline tests/performance_tests/performance_test.py:9]: Can we group these between imports and froms?

**Review by ClaytonCalabrese — APPROVED** [review]
Looks fine, just one formatting complaint. Running locally now to verify works as intended.

File: tests/performance_tests/performance_test.py (hunk 2 above)
Comment (oschwaldp-oci) [reply tests/performance_tests/performance_test.py]: Done. https://github.com/AntelopeIO/leap/pull/302/commits/7e9117ec2f4e47755343bd6177e335f57ea9c32c

**Review by heifner — APPROVED** [review]

**Review by jgiszczak — CHANGES_REQUESTED** [review]

File: tests/performance_tests/performance_test.py [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+    js['nodeosVersion'] = Utils.getNodeosVersion()
+    return json.dumps(js, indent=2)
+
+def exportReportAsJSON(report: json, exportPath):
+    with open(exportPath, 'wt') as f:
+        f.write(report)
+
+def testDirsCleanup(saveJsonReport, testTimeStampDirPath, ptbLogsDirPath):
+    try:
+        if saveJsonReport:
```
[the remark this answers is not in the document]
Comment (oschwaldp-oci) [reply tests/performance_tests/performance_test.py:126]: Please see: [d45d5b0](https://github.com/AntelopeIO/leap/pull/302/commits/d45d5b0551c0f8d3f26dcaf740ff87f4ba0724a2)

File: tests/performance_tests/performance_test.py [hunk 4]
```diff
@@ -0,0 +1,271 @@
...
+
+    except OSError as error:
+        print(error)
+
+def prepArgsDict(testDurationSec, finalDurationSec, testTimeStampDirPath, maxTpsToTest, testIterationMinStep,
+             tpsLimitPerGenerator, saveJsonReport, saveTestJsonReports, numAddlBlocksToPrune, testHelperConfig, testClusterConfig) -> dict:
+    argsDict = {}
+    argsDict.update(asdict(testHelperConfig))
+    argsDict.update(asdict(testClusterConfig))
+    argsDict["testDurationSec"] = testDurationSec
```
[the remark this answers is not in the document]
Comment (oschwaldp-oci) [reply tests/performance_tests/performance_test.py]: Please see: [dab4422](https://github.com/AntelopeIO/leap/pull/302/commits/dab4422fa07125d7024e48733f9858967d4702e0)

File: tests/performance_tests/performance_test.py [hunk 5]
```diff
@@ -0,0 +1,271 @@
...
+            maxTpsReport = json.loads(myTest.report)
+            floor = binSearchTarget
+            scenarioResult.success = True
+        else:
+            ceiling = binSearchTarget
+
+        scenarioResult.basicTestResult = ptbResult
+        searchResults.append(scenarioResult)
+        print(f"searchResult: {binSearchTarget} : {searchResults[-1]}")
+        if lastRun:
```
[the remark this answers is not in the document]
Comment (oschwaldp-oci) [reply tests/performance_tests/performance_test.py]: Please see: [97711cb](https://github.com/AntelopeIO/leap/pull/302/commits/97711cb482572e5dc6f930ed04eb604e0d1a483e)

**Review by jgiszczak — APPROVED** [review]

**Pull request merged**